cli/src/plz/cli/retrieve_output_operation.py

```diff
@@ -0,0 +1,94 @@
+import io
+import os
+import shutil
+import tarfile
+import tempfile
+from typing import Iterator, Optional
+
+import requests
+
+from plz.cli.configuration import Configuration
+from plz.cli.exceptions import CLIException
+from plz.cli.log import log_info
+from plz.cli.operation import Operation, check_status, on_exception_reraise
+from plz.cli.show_status_operation import ShowStatusOperation
+
+
+class RetrieveOutputOperation(Operation):
+    @staticmethod
+    def prepare_argument_parser(parser, args):
+        if len(args) > 1:
+            # Execution ID was specified
+            parser.add_argument(dest='execution_id')
+        cwd = os.getcwd()
+        parser.add_argument('-o', '--output-dir',
+                            type=str,
+                            default=os.path.join(cwd, 'output'))
+
+    def __init__(self, configuration: Configuration,
+                 output_dir: str,
+                 execution_id: Optional[str] = None):
+        super().__init__(configuration)
+        self.output_dir = output_dir
+        self.execution_id = execution_id
+
+    def harvest(self):
+        log_info('Harvesting the output...')
+        response = requests.delete(
+            self.url('executions', self.get_execution_id()))
+        check_status(response, requests.codes.no_content)
+
+    @on_exception_reraise('Retrieving the output failed.')
+    def retrieve_output(self):
+        execution_id = self.get_execution_id()
+        log_info('Retrieving the output...')
+        response = requests.get(
+            self.url('executions', execution_id, 'output', 'files'),
+            stream=True)
+        check_status(response, requests.codes.ok)
+        try:
+            os.makedirs(self.output_dir)
+        except FileExistsError:
+            raise CLIException(
+                f'The output directory "{self.output_dir}" already exists.')
+        for path in untar(response.raw, self.output_dir):
+            print(path)
+
+    def run(self):
+        show_status_operation = ShowStatusOperation(
+            self.configuration, self.get_execution_id())
+        if show_status_operation.get_status().running:
+            raise CLIException(
+                'Process is still running. Use `plz stop` if '
+                'you want to terminate it')
```

Comment:
In the spirit of Python's mentality, favouring exceptions over up-front checks, should we have `.harvest()` throw an exception the process is still running? (And have the `DELETE` endpoint respond with a *409 Conflict*.)

We could then catch the exception and re-raise, rather than having to do a separate query.

Basically, moving logic to the server also makes things more Pythonic, and I like both these things.

Reply:
Oh, sorry, I misread it for other code we have checking whether the process has finished after control returned to the cli (in attached mode). Delete currently just deletes no matter what, and I think we need to keep that way (I was thinking that's the entrypoint we want to hit with `plz stop` so that it's consistent with the attached mode --instead of the current stop entrypoint, that doesn't trigger a harvest). I can add an argument to the argument entrypoint as to fail_if_running.

Reply:
Yeah, I think there's probably too many endpoints doing similar things. Leave it as is for now, then, and let's clean them up later, probably as part of making harvesting a background job.